src/main/java/org/prebid/server/bidder/huaweiads/HuaweiAdSlotBuilder.java

```diff
@@ -139,15 +156,76 @@ private List<Asset> parseNativeRequestAssets(Native xNative) {
     }
 
     private static List<String> makeDetailedCreativeTypeList(long numVideo, long numImage) {
+        final List<String> detailedCreativeTypeList = new ArrayList<>();
         if (numVideo >= 1) {
-            return List.of("903");
-        } else if (numImage > 1) {
-            return List.of("904");
-        } else if (numImage == 1) {
-            return List.of("901");
-        } else {
-            return List.of("913", "914");
+            detailedCreativeTypeList.add("903");
         }
+        if (numImage >= 1) {
+            detailedCreativeTypeList.addAll(List.of("901", "904", "905"));
+        }
+        return detailedCreativeTypeList;
+    }
+
+    private static List<Format> makeFormatListForNative(List<Asset> assets, long numImage) {
+        final Set<Format> formats = new HashSet<>();
+        formats.addAll(makeFormatListForVideo(assets));
+        formats.addAll(makeFormatListForImage(assets, numImage));
+        return formats.stream().toList();
+    }
+
+    private static Set<Format> makeFormatListForVideo(List<Asset> assets) {
+        return assets.stream()
+                .map(Asset::getVideo)
+                .filter(Objects::nonNull)
+                .filter(video -> HuaweiUtils.isFormatDefined(video.getW(), video.getH()))
+                .map(video -> Format.of(video.getW(), video.getH()))
+                .collect(Collectors.toSet());
     }
 
+    private static Set<Format> makeFormatListForImage(List<Asset> assets, long numImage) {
+        final Set<Format> formats = new HashSet<>();
+        assets.stream()
+                .map(Asset::getImg)
+                .filter(Objects::nonNull)
+                .forEach(image -> {
+                    if (numImage > 1
+                            && HuaweiUtils.isFormatDefined(image.getW(), image.getH())
+                            && HuaweiUtils.isFormatDefined(image.getWmin(), image.getHmin())) {
+                        formats.add(Format.of(image.getW(), image.getH()));
+                    }
+                    if (numImage == 1
+                            && HuaweiUtils.isFormatDefined(image.getW(), image.getH())
+                            && HuaweiUtils.isFormatDefined(image.getWmin(), image.getHmin())) {
+                        formats.addAll(filterPopularSizesByRatio(image.getW(), image.getH()));
+                    }
+                    if (numImage == 1
+                            && !HuaweiUtils.isFormatDefined(image.getW(), image.getH())
+                            && HuaweiUtils.isFormatDefined(image.getWmin(), image.getHmin())) {
+                        formats.addAll(filterPopularSizesByRange(image.getWmin(), image.getHmin()));
```

Comment:
I still think those two `isFormatDefined` calls are worth calculating once per image, but not three times

P.S. Moving `foreach` logic into a separate method still makes sense
P.S.S. Instead of the `foreach` it's possible to use `flatmap` in order to get the list of filtered format and merge them into one set, for example

```
private static Set<Format> makeFormatListForImage(List<Asset> assets, long numImage) {
       return assets.stream()
                .map(Asset::getImg)
                .filter(Objects::nonNull)
                .map(image -> filterFormats(image, numImage))
                .flatMap(Collection::stream)
                .collect(Collectors.toSet());
    }
 ```

Reply:
Fixed.